### PR TITLE
Feature/exam mode/exam management

### PR DESCRIPTION
**Pull request by anditurdiu**
# Checklist
- [x] I tested all changes and all related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
This PR creates the exam management view and the functionality to display exams for courses as well as delete exams. See #1618

### Description
Created a view for the exam management and forwarded the appropriate requests to the server. 
Within the scope of this PR are the listing of the exams and the ability to delete an exam from the server. For all other buttons, @sascha11110 's routes have been used but they link to 'TODO' sites or throw an error if there is no appropriate mapping yet. 

### Steps for Testing

1. Go to https://artemistest.ase.in.tum.de
2. Log in to Artemis
3. Go to Course Management
4. Click on Exams
5. Create an Exam.
6. The Exam should now be visible if you are an admin/instructor
7. Delete a Course, it should be deleted from the server

### Screenshots

- Course Management
image

- Exam Management
image

- Delete Exam 
image

### CODE REVIEW

File: src/main/webapp/app/course/manage/course-management.component.html [hunk 1]
```diff
@@ -111,18 +111,6 @@ <h4 id="course-page-heading" jhiTranslate="artemisApp.course.home.title">Courses
                             <span *ngIf="course.presentationScore !== 0">{{ 'global.generic.yes' | translate }}</span>
                             <span *ngIf="course.presentationScore === 0">{{ 'global.generic.no' | translate }}</span>
                         </td>
-                        <td jhiOrionFilter [showInOrionWindow]="true">
```
Comment (anditurdiu) [inline src/main/webapp/app/course/manage/course-management.component.html:114]: I think this was an accidental commit a while ago as this button is actually present within the 
`<div class="btn-group flex-btn-group-container">`

File: src/main/webapp/i18n/de/exam.json [hunk 2]
```diff
@@ -1,10 +1,23 @@
 {
     "artemisApp": {
         "exam": {
-            "title": "Exam"
+            "title": "Klausur",
+            "visible": "Sichtbar",
+            "startDate": "Anfang",
```
Comment (julian-christl) [inline src/main/webapp/i18n/de/exam.json:6]: Wouldn't be "Beginn" be a better naming? "Beginn" indicates a point in time whereas "Anfang" indicates more an area in time imo.

File: src/main/webapp/app/exam/manage/exam-management.component.html [hunk 3]
```diff
@@ -1 +1,119 @@
-TODO
+<div>
+    <div class="d-flex">
+        <h4 id="course-page-heading" jhiTranslate="artemisApp.examManagement.title">Exams</h4>
+        <div class="ml-auto d-flex">
+            <button jhiOrionFilter class="btn btn-primary jh-create-entity create-exam" *jhiHasAnyAuthority="['ROLE_ADMIN', 'ROLE_INSTRUCTOR']" [routerLink]="['new']">
```
Comment (sascha11110) [inline src/main/webapp/app/exam/manage/exam-management.component.html:5]: More a question as an comment since I stumbled across it too: Do we need `jhiOrionFilter` and corresponding attributes for the exam management?

File: src/main/webapp/app/shared/pipes/artemis-duration.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        startDate: Date | moment.Moment | string | number | null,
+        endDate: Date | moment.Moment | string | number | null,
+        format: 'short' | 'long' | 'short-date' | 'long-date' | 'time' = 'long',
+    ): string {
+        if (!endDate || !moment(endDate).isValid() || !startDate || !moment(startDate).isValid()) {
+            return '';
+        }
+        if (moment(endDate).diff(startDate, 'hours') < 24) {
+            console.log(moment(endDate).diff(startDate, 'hours'));
+            return moment.utc(moment(endDate, 'DD/MM/YYYY HH:mm:ss').diff(moment(startDate, 'DD/MM/YYYY HH:mm'))).format('H:mm');
```
Comment (sascha11110) [inline src/main/webapp/app/shared/pipes/artemis-duration.pipe.ts:23]: `HH:mm` is used in `ArtemisDatePipe` so it would be good to be consistent here :-)

File: src/main/webapp/app/shared/pipes/artemis-duration.pipe.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    transform(
+        startDate: Date | moment.Moment | string | number | null,
+        endDate: Date | moment.Moment | string | number | null,
+        format: 'short' | 'long' | 'short-date' | 'long-date' | 'time' = 'long',
+    ): string {
+        if (!endDate || !moment(endDate).isValid() || !startDate || !moment(startDate).isValid()) {
+            return '';
+        }
+        if (moment(endDate).diff(startDate, 'hours') < 24) {
+            console.log(moment(endDate).diff(startDate, 'hours'));
```
Comment (sascha11110) [inline src/main/webapp/app/shared/pipes/artemis-duration.pipe.ts:22]: Remove ;-)

File: src/main/webapp/app/shared/pipes/artemis-duration.pipe.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+            return moment.utc(moment(endDate, 'DD/MM/YYYY HH:mm:ss').diff(moment(startDate, 'DD/MM/YYYY HH:mm'))).format('H:mm');
+        } else {
+            console.log(moment(endDate).diff(startDate, 'hours'));
```
Comment (sascha11110) [inline src/main/webapp/app/shared/pipes/artemis-duration.pipe.ts:25]: Remove ;-)

File: src/main/webapp/app/shared/pipes/artemis-duration.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+     * Calculate the duration of an event from the diff of two dates.
+     * @param startDate The start date time of the event. Must be convertible to moment().
+     * @param endDate The end date time of the event. Must be convertible to moment().
+     * @param format Format of the localized date time. Defaults to 'long'.
+     * @returns string 'HH:mm'
+     */
+    transform(
+        startDate: Date | moment.Moment | string | number | null,
+        endDate: Date | moment.Moment | string | number | null,
+        format: 'short' | 'long' | 'short-date' | 'long-date' | 'time' = 'long',
```
Comment (sascha11110) [inline src/main/webapp/app/shared/pipes/artemis-duration.pipe.ts:16]: You are not using it so I think you can remove it. Also `ArtemisDatePipe` does not apply the format *length* to its time format so you will definitely not need it.

File: src/main/webapp/app/exam/manage/exam-management.module.ts [hunk 8]
```diff
@@ -11,11 +11,16 @@ import { ExerciseGroupDetailComponent } from 'app/exam/manage/exercise-groups/ex
...
+import { SortByModule } from 'app/shared/pipes/sort-by.module';
+import { ArtemisSharedModule } from 'app/shared/shared.module';
+import { OrionModule } from 'app/shared/orion/orion.module';
+import { DurationPipe } from 'app/shared/pipes/artemis-duration.pipe';
 
 const ENTITY_STATES = [...examManagementState];
 
 @NgModule({
-    imports: [RouterModule.forChild(ENTITY_STATES)],
+    imports: [RouterModule.forChild(ENTITY_STATES), ArtemisSharedModule, OrionModule, SortByModule, MomentModule],
```
Comment (sascha11110) [inline src/main/webapp/app/exam/manage/exam-management.module.ts:23]: Again: do wee need `OrionModule` here?

File: src/main/webapp/app/exam/manage/exam-management.component.html (hunk 3 above)
Comment (anditurdiu) [reply src/main/webapp/app/exam/manage/exam-management.component.html:5]: I'm not sure, i added it as it was present in the other similar components such as the course-management component

**Comment by pzdr7** [comment]
I've tested this locally with multiple exams and everything seems to work as expected. (Obviously, exam creation, editing and the assessment dashboard do not work yet because they are not a part of this branch).

File: src/main/webapp/app/course/manage/course-management.component.ts [hunk 9]
```diff
@@ -20,7 +20,6 @@ export class CourseManagementComponent implements OnInit, OnDestroy, AfterViewIn
     predicate: string;
     reverse: boolean;
     showOnlyActive = true;
-    showExamButton = false;
```
Comment (krusche) [inline src/main/webapp/app/course/manage/course-management.component.ts:23]: I would like to keep this feature disabled on production for now. So we should keep this attribute.
You can set it to true for testing, but we should set it to false again before we deploy the new features to production

File: src/main/webapp/app/course/manage/course-management.component.html [hunk 10]
```diff
@@ -149,7 +137,7 @@ <h4 id="course-page-heading" jhiTranslate="artemisApp.course.home.title">Courses
                                 </div>
                                 <div class="btn-group-vertical mr-1 mb-1" *jhiHasAnyAuthority="['ROLE_ADMIN', 'ROLE_INSTRUCTOR']">
                                     <button
-                                        *ngIf="showExamButton && course.isAtLeastInstructor"
+                                        *ngIf="course.isAtLeastInstructor"
```
Comment (krusche) [inline src/main/webapp/app/course/manage/course-management.component.html:140]: please revert this change

File: src/main/webapp/app/exam/manage/exam-management.component.html (hunk 3 above)
Comment (krusche) [reply src/main/webapp/app/exam/manage/exam-management.component.html:5]: This is a specific filter for Orion. It specifies that the button should be shown in Orion. I think for now we can leave this out for all exam related functionalities

File: src/main/webapp/app/course/manage/course-management.component.html (hunk 1 above)
Comment (krusche) [reply src/main/webapp/app/course/manage/course-management.component.html:114]: please revert this change, this is a specific distinction for Orion, the IntelliJ plugin. I guess this button is shown in Orion (not in Artemis), the same one (below) is shown in Artemis, but not in Orion.

File: src/main/webapp/app/exam/manage/exam-management.module.ts (hunk 8 above)
Comment (krusche) [reply src/main/webapp/app/exam/manage/exam-management.module.ts:23]: I would leave it out for now

**Comment by julian-christl** [comment]
If we merge the develop in here it should be testable in the UI.

**Comment by sascha11110** [comment]
I've resolved the merge conflicts and addressed @krusche's comments.

**Just that we don't forget:**
We should set `showExamButton` in `course-management.component.ts` to `false` before we merge.

File: src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java [hunk 11]
```diff
@@ -243,4 +243,10 @@ public void testDeleteExam_asInstructor() throws Exception {
         request.delete("/api/courses/" + course1.getId() + "/exams/" + exam1.getId(), HttpStatus.OK);
         verify(examAccessService, times(1)).checkCourseAndExamAccess(course1.getId(), exam1.getId());
     }
+
+    @Test
+    @WithMockUser(value = "admin", roles = "ADMIN")
+    public void DeleteExamThatDoesNotExist() throws Exception {
+        request.delete("/api/courses/" + courseRepo.findAll().get(0).getId() + "/exams/1", HttpStatus.NOT_FOUND);
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java:250]: Why not chossing a specific course? Like course1 or course2 from line 75 and 76.

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java [hunk 12]
```diff
@@ -516,6 +516,41 @@ public Exam addExam(Course course) {
         return exam;
     }
 
+    public Exam addExamToCourse(Course course) {
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java:519]: What is the benefit of defining this here? (In comparison to the addExam() one above using ModelFactory.generateExam(course))

Comment (anditurdiu) [reply src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java]: This was actually unused code as the exam generation was refactored in another branch which was merged into develop. Good catch! I removed it

File: src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java (hunk 11 above)
Comment (anditurdiu) [reply src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java]: No real reason, for consistency i replaced the hardcoded value and am now using the id of the specified course.

**Comment by anditurdiu** [comment]
> Not sure if it's just a bug on my side but when I try to click on "Exams" to create an exam, nothing happens. I changed the accounts (admin to instructor) but still the same behaviour.

I cannot reproduce this behaviour. After creating an exam it appears on the list. Can you further elaborate on how this happened?

**Comment by amahmario** [comment]
> > Not sure if it's just a bug on my side but when I try to click on "Exams" to create an exam, nothing happens. I changed the accounts (admin to instructor) but still the same behaviour.
> 
> I cannot reproduce this behaviour. After creating an exam it appears on the list. Can you further elaborate on how this happened?

Nevermind it works for me now! 👍
